src/Core/MySQL/PacketsProtocolBinary.cpp

```diff
@@ -0,0 +1,363 @@
+                    }
+                    case TypeIndex::Date32: {
+                        payload_size += 5;
+                        break;
+                    }
+                    case TypeIndex::DateTime: {
+                        UInt32 value = assert_cast<const ColumnVector<UInt32> &>(*col).getData()[row_num];
+                        LocalDateTime ldt = LocalDateTime(value, DateLUT::instance(getDateTimeTimezone(*data_type)));
+
+                        bool has_time = !(ldt.hour() == 0 && ldt.minute() == 0 && ldt.second() == 0);
+                        if (has_time)
+                        {
+                            payload_size += 8;
+                        }
+                        else
+                        {
+                            payload_size += 5;
+                        }
+                        break;
+                    }
+                    case TypeIndex::DateTime64: {
+                        const auto * date_time_type = typeid_cast<const DataTypeDateTime64 *>(data_type.get());
+                        UInt32 scale = date_time_type->getScale();
+
+                        static constexpr UInt32 MaxScale = DecimalUtils::max_precision<DateTime64>;
+                        scale = scale > MaxScale ? MaxScale : scale;
+
+                        const auto dt64 = assert_cast<const ColumnDateTime64 &>(*col).getData()[row_num];
+                        auto components = DecimalUtils::split(dt64, scale);
+
+                        using T = typename DateTime64::NativeType;
+                        if (dt64.value < 0 && components.fractional)
+                        {
+                            components.fractional
+                                = DecimalUtils::scaleMultiplier<T>(scale) + (components.whole ? T(-1) : T(1)) * components.fractional;
+                            --components.whole;
+                        }
+                        if (scale > 6)
+                        {
+                            // MySQL Timestamp has max scale of 6
+                            components.fractional /= static_cast<int>(pow(10, scale - 6));
+                        }
+
+                        LocalDateTime ldt = LocalDateTime(components.whole, DateLUT::instance(getDateTimeTimezone(*data_type)));
+
+                        bool has_microseconds = components.fractional != 0;
+                        bool has_time = !(ldt.hour() == 0 && ldt.minute() == 0 && ldt.second() == 0);
+                        if (has_microseconds)
+                        {
+                            payload_size += 12;
+                        }
+                        else if (has_time)
+                        {
+                            payload_size += 8;
+                        }
+                        else
+                        {
+                            payload_size += 5;
+                        }
+                        break;
+                    }
+                    default:
```

Comment:
Should there be a handler for Decimal?

Reply:
Actually, no, cause it is just a string(lenenc) handled by the default case. Maybe I should add a comment about that.